### PR TITLE
[OM-97044]: ORM v2 discovery

### PR DESCRIPTION
**Pull request by pallavidn**
New version of Operator Resource Mapping (ORM) resource is now available via [Turbonomic ORM project](https://github.com/turbonomic/orm).

K8s probe discovery is updated to discover this new version. 
Subsequent changes will be made to the K8s action execution to resize controller actions that are controlled by operators using the new version of ORMs.

Summary of changes in this PR:

- This large commit due to vendor update changes. ORM project is required to work with the ORM v2 data structures and the registry structure to save the discovered ORMs.
- Provided an interface to handle the both the legacy and v2 ORMs 
- This interface is now available to the discovery and action executor clients of the k8s probe, in lieu of the previous ORM client.

Process of discovery of OMRs is as follows:
-  Create a controller client to fetch the list of OperatorResourceMapping resources from the k8s cluster
-  Use the ResourceMappingRegistry from the orm project to validate and register the ORMs. Validation step checks that the owner object (which controls the workload controllers that are discovered by the kubeturbo probe) exists. Registry interface provides a convenient access to the all the  resource paths mapping defined using parameters in the ORM.
Example in the ORM defined below, the owner path pointing to the source resources is parameterized.
```
spec:
  owner: # target same namespace by default
    apiVersion: apps/v1
    kind: Deployment
    name: ormoperand-patterns
  mappings:
    parameters:
      ports:
      - "TCP"
      - "UDP"
    patterns:
    - ownerPath: .spec.template.spec.containers[?(@.name=="{{.owned.name}}")].ports[?(@.protocol=="{{ports}}")].containerPort
      owned:
        apiVersion: apps/v1
        kind: Deployment
        matchLabels:
          app: ormsource-patterns
        path: .spec.template.spec.containers[?(@.name=="{{.owned.name}}")].ports[?(@.protocol=="{{ports}}")].containerPort

```

-  Set the status in the ORM with the current state of the owner object (resource paths and resource values). Status will update the ORM object with the owner object reference and list of the expanded resource paths (corresponding the resource in the source objects) with their current value.
```
status:
  owner:
    apiVersion: apps/v1
    kind: Deployment
    name: ormoperand-patterns
    namespace: turbo-et
  ownerValues:
  - owned:
      apiVersion: apps/v1
      kind: Deployment
      name: ormsource-patterns-0001
      namespace: turbo-et
      path: .spec.template.spec.containers[?(@.name=="ormsource-patterns-0001")].ports[?(@.protocol=="UDP")].containerPort
    ownerPath: .spec.template.spec.containers[?(@.name=="ormsource-patterns-0001")].ports[?(@.protocol=="UDP")].containerPort
    value:
      containerPort: 10001
  - owned:
      apiVersion: apps/v1
      kind: Deployment
      name: ormsource-patterns-0002
      namespace: turbo-et
      path: .spec.template.spec.containers[?(@.name=="ormsource-patterns-0002")].ports[?(@.protocol=="TCP")].containerPort
    ownerPath: .spec.template.spec.containers[?(@.name=="ormsource-patterns-0002")].ports[?(@.protocol=="TCP")].containerPort
    value:
      containerPort: 82
  - owned:
      apiVersion: apps/v1
      kind: Deployment
      name: ormsource-patterns-0002
      namespace: turbo-et
      path: .spec.template.spec.containers[?(@.name=="ormsource-patterns-0002")].ports[?(@.protocol=="UDP")].containerPort
    ownerPath: .spec.template.spec.containers[?(@.name=="ormsource-patterns-0002")].ports[?(@.protocol=="UDP")].containerPort
    value:
      containerPort: 10002
  - owned:
      apiVersion: apps/v1
      kind: Deployment
      name: ormsource-patterns-0001
      namespace: turbo-et
      path: .spec.template.spec.containers[?(@.name=="ormsource-patterns-0001")].ports[?(@.protocol=="TCP")].containerPort
    ownerPath: .spec.template.spec.containers[?(@.name=="ormsource-patterns-0001")].ports[?(@.protocol=="TCP")].containerPort
    value:
      containerPort: 81
  state: ok

```

For testing reference, see the ORM Status 
- When owner or resource paths are invalid

```
status:
  owner:
    apiVersion: apps/v1
    kind: Deployment
    name: ormowner-invalid
    namespace: turbo-et
  ownerValues:
  - message: Failed to locate mapping path .spec.invalid.replicas in owned resource
    owned:
      apiVersion: apps/v1
      kind: Deployment
      name: ormowned-invalid
      namespace: turbo-et
      path: .spec.invalid.replicas
    ownerPath: .spec.replicas
    reason: OwnedResourceError
    value:
      replicas: 3
  - message: Failed to locate ownerPath in owner
    ownerPath: .spec.template.spec.containers[?(@.name=="workload-0001")].invalid.resources
    reason: OwnerError
  state: ok

```

- When the source object is not found
```
status:
  owner:
    apiVersion: apps/v1
    kind: Deployment
    name: ormowner-missing-owned
    namespace: turbo-et
  ownerValues:
  - message: 'Failed to locate owned resource: &ObjectReference{Kind:Deployment,Namespace:turbo-et,Name:ormowned-missing-owned,UID:,APIVersion:apps/v1,ResourceVersion:,FieldPath:,}'
    owned:
      apiVersion: apps/v1
      kind: Deployment
      name: ormowned-missing-owned
      namespace: turbo-et
      path: .spec.replicas
    ownerPath: .spec.replicas
    reason: OwnedResourceError
    value:
      replicas: 3
  - message: 'Failed to locate owned resource: &ObjectReference{Kind:Deployment,Namespace:turbo-et,Name:ormowned-missing-owned,UID:,APIVersion:apps/v1,ResourceVersion:,FieldPath:,}'
    owned:
      apiVersion: apps/v1
      kind: Deployment
      name: ormowned-missing-owned
      namespace: turbo-et
      path: .spec.template.spec.containers[?(@.name=="workload-0001")].resources
    ownerPath: .spec.template.spec.containers[?(@.name=="workload-0001")].resources
    reason: OwnedResourceError
    value:
      resources:
        limits:
          cpu: 100m
          memory: 100Mi
        requests:
          cpu: 50m
          memory: 50Mi
  state: ok 
```

- When the owner object is not found
ORM is not registered

### CODE REVIEW

**Comment by kuanf** [comment]
btw, I think the status should include the owned object reference as well, not just ownerPath and value. This is to satisfy requirement from cp4i to use the mapping themselves.  
Did you apply the very latest CRD to your system (again)? :) 

```
  status:
    owner:
      apiVersion: apps/v1
      kind: Deployment
      name: ormoperand-patterns
      namespace: default
    ownerValues:
    - owned:
        apiVersion: apps/v1
        kind: Deployment
        name: ormsource-patterns-0001
        namespace: default
        path: .spec.template.spec.containers[?(@.name=="ormsource-patterns-0001")].ports[?(@.protocol=="TCP")].containerPort
      ownerPath: .spec.template.spec.containers[?(@.name=="ormsource-patterns-0001")].ports[?(@.protocol=="TCP")].containerPort
      value:
        containerPort: 81
    - owned:
        apiVersion: apps/v1
        kind: Deployment
        name: ormsource-patterns-0001
        namespace: default
        path: .spec.template.spec.containers[?(@.name=="ormsource-patterns-0001")].ports[?(@.protocol=="UDP")].containerPort
      ownerPath: .spec.template.spec.containers[?(@.name=="ormsource-patterns-0001")].ports[?(@.protocol=="UDP")].containerPort
      value:
        containerPort: 10001
    - owned:
        apiVersion: apps/v1
        kind: Deployment
        name: ormsource-patterns-0002
        namespace: default
        path: .spec.template.spec.containers[?(@.name=="ormsource-patterns-0002")].ports[?(@.protocol=="TCP")].containerPort
      ownerPath: .spec.template.spec.containers[?(@.name=="ormsource-patterns-0002")].ports[?(@.protocol=="TCP")].containerPort
      value:
        containerPort: 82
    - owned:
        apiVersion: apps/v1
        kind: Deployment
        name: ormsource-patterns-0002
        namespace: default
        path: .spec.template.spec.containers[?(@.name=="ormsource-patterns-0002")].ports[?(@.protocol=="UDP")].containerPort
      ownerPath: .spec.template.spec.containers[?(@.name=="ormsource-patterns-0002")].ports[?(@.protocol=="UDP")].containerPort
      value:
        containerPort: 10002
    state: ok
```

File: pkg/resourcemapping/v2/orm_client_v2.go [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+
+	// Register ORM if owner exists
+	err = ormClient.ValidateAndRegisterORM(orm)
+	if err != nil {
+		return err
+	}
+	glog.V(3).Infof("Registered orm %v:%v", orm.Namespace, orm.Name)
+
+	var ownerObj *unstructured.Unstructured
+	ownerObj, err = ormClient.getOwnerFromORM(orm)
```
[the remark this answers is not in the document]
Comment (kuanf) [reply pkg/resourcemapping/v2/orm_client_v2.go:100]: created the pr to orm repo. https://github.com/turbonomic/orm/pull/22

**Comment by pallavidn** [comment]
Updated observed ORM status values  in the PR description with the owner and owned references.

File: pkg/resourcemapping/v2/orm_client_v2.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (pallavidn) [reply pkg/resourcemapping/v2/orm_client_v2.go:100]: Yes, orm#22 approved.

**Review by kuanf — APPROVED** [review]

**Pull request merged**